### DISCUSSION TITLE
How to read environment variables in a js file in the frontend

### DISCUSSION THREAD

**Discussion by rafakwolf** [11ty/eleventy#2489] [Uncategorized]
Here's the thing.
I need to make API calls to a certain service, and the service requires an API token.
So I do not want to expose the API token in the front-end code.
I was attempting to use environment vars within a data file, but how can I read it inside a javascript file in the browser?
What would be the best way to do that?

using 11ty 0.9.0

Thank you!

**Comment by pdehaan** [comment]
Sorry, to clarify, are you trying to read an environment variable within a global _data/foo.js data file at build time?
Or are you saying your static site is calling an API at runtime via a deployed .js file?

**Comment by rafakwolf** [comment]
Hey @pdehaan, thank you for the answer.

> Or are you saying your static site is calling an API at runtime via a deployed .js file?

Correct ^

**Reply by pdehaan** [reply]
Ah, in that case I might look at https://www.11ty.dev/docs/plugins/serverless/ or possibly some other serverless Node script on GCP/AWS/Netlify/Vercel/etc that you can proxy through (and potentially cache, if the requests aren't unique to the user).
At least that way you can query your endpoint and keep your API key hidden.
